ykrt/src/mt.rs

```diff
@@ -332,17 +347,38 @@ impl MTThread {
                 }
             }
             PHASE_TRACING => {
-                if let Some((_, tid)) = self.inner.tracer {
+                let loc_tid = unsafe { Arc::from_raw((lp & !PHASE_TAG) as *mut u8) };
+                if let Some((_, ref tid)) = self.inner.tracer {
                     // This thread is tracing something...
-                    if tid != ((lp & !PHASE_TAG) as *mut u8) {
+                    if !Arc::ptr_eq(tid, &loc_tid) {
                         // ...but we didn't start at the current Location.
+                        mem::forget(loc_tid);
                         return None;
                     } else {
                         // ...and we started at this Location, so we've got a complete loop!
                     }
                 } else {
                     // Another thread is tracing this location.
-                    return None;
+                    if Arc::strong_count(&loc_tid) == 1 {
+                        // The other thread has stopped. Mark this Location as WONT_TRACE.
+                        if loc
+                            .state
+                            .compare_and_swap(lp, PHASE_WONT_TRACE, Ordering::Acquire)
```

Comment:
So this can be a regular store, if we know all racing threads will store the same value `PHASE_WONT_TRACE`?

Reply:
No, because there could be another live (but non-tracing thread) that's carrying out the same check as us, so we might race on the transition to `PHASE_WONT_TRACE`. If we did race, we'd both cause the `Arc` to be `free`d (i.e. a double free).

Reply:
Gotcha. Nice catch! I'd have screwed that up!